src/gen_rules.ml

```diff
@@ -520,6 +522,49 @@ module Gen(P : Params) = struct
          >>>
          Build.and_create_file digest_path)
 
+  let copy_files_rules (def: Copy_files.t) ~src_dir ~dir ~package_context:_ =
+    let src_glob = SC.expand_vars sctx ~dir def.glob in
+    let glob_in_src = Path.relative src_dir src_glob in
+    let glob = Path.basename glob_in_src in
+    let src_in_src = Path.parent glob_in_src in
+    let re = match Glob_lexer.parse_string glob with
+    | Ok re ->
+      Re.compile re
+    | Error (_pos, msg) ->
+      die "invalid glob in %s/jbuild: %s" (Path.to_string src_dir) msg
+    in
+    (* add rules *)
+    let modules = SC.sources_and_targets_known_so_far sctx ~src_path:src_in_src in
+    let src_in_build = Path.append ctx.build_dir src_in_src in
+    String_set.iter modules ~f:(fun basename ->
+      let matches = Re.execp re basename in
+      if matches then
+        let file_src = Path.relative src_in_build basename in
+        let file_dst = Path.relative dir basename in
+        SC.add_rule sctx
+          (Build.path file_src
+           >>>
+           let action =
+             if def.add_line_directive
+             then Action.Mini_shexp.Ast.Copy_and_add_line_directive
+                    (file_src,file_dst)
+             else Action.Mini_shexp.Ast.Copy (file_src,file_dst)
+           in
+           (Build.action 
```

Comment:
There is already `Build.copy`. Let's add Build.copy_and_add_line_directive as well. This will make this code less verbose.